controllers/storageclassclaim/storageclassclaim_controller.go

```diff
@@ -92,14 +121,241 @@ func (r *StorageClassClaimReconciler) Reconcile(ctx context.Context, request rec
 	return result, nil
 }
 
-func (r *StorageClassClaimReconciler) reconcileConsumerPhases(ctx context.Context, storagecluster *v1.StorageCluster) (reconcile.Result, error) {
-	log := ctrllog.FromContext(ctx)
-	log.Info("Running StorageClassClaim controller in Consumer Mode")
+func (r *StorageClassClaimReconciler) reconcileConsumerPhases() (reconcile.Result, error) {
+	r.log.Info("Running StorageClassClaim controller in Consumer Mode")
+
+	providerClient, err := providerclient.NewProviderClient(
+		r.ctx,
+		r.storageCluster.Spec.ExternalStorage.StorageProviderEndpoint,
+		10*time.Second,
+	)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Close client-side connections.
+	defer providerClient.Close()
+
+	if r.storageClassClaim.GetDeletionTimestamp().IsZero() {
+
+		// Check if finalizers are present, if not, add them.
+		if !contains(r.storageClassClaim.GetFinalizers(), StorageClassClaimFinalizer) {
+			storageClassClaimRef := klog.KRef(r.storageClassClaim.Name, r.storageClassClaim.Namespace)
+			r.log.Info("Finalizer not found for StorageClassClaim. Adding finalizer.", "StorageClassClaim", storageClassClaimRef)
+			r.storageClassClaim.SetFinalizers(append(r.storageClassClaim.GetFinalizers(), StorageClassClaimFinalizer))
+			if err := r.update(r.storageClassClaim); err != nil {
+				return reconcile.Result{}, fmt.Errorf("failed to update StorageClassClaim [%v] with finalizer: %s", storageClassClaimRef, err)
+			}
+		}
+
+		// Creation phase.
+		if r.storageClassClaim.Status.Phase == "" {
```

Comment:
This will not work as if fulfill fails with an error the next reconciliation will skip this section because the resource will be in Creating phase and you will never fulfill the claim. 

Please add an indicator in the status section of the CR to indicate whether the claim is fulfilled or not and skip this section only if the indicator is true. The indicator should be separate from the value of the `Phase` field as the `Phase` field can lose its value if we fail on a prior phase, for example, the `Validating` phase that is missing from this commit

Reply:
> This will not work as if fulfill fails with an error the next reconciliation will skip this section because the resource will be in Creating phase and you will never fulfill the claim.

PTAL https://github.com/red-hat-storage/ocs-operator/pull/1608/files#diff-b60803ce0dfb3bc9a7a20db01eec72ffd2803fc90febca9acd9a0cea7af477dfR173-R174

Reply:
Resetting to `""` will not solve the issue, you will just lose your position in the recent reconciliation (as `State` is always committed at the end of the reconciliation), and even if not the phase and fulfillment states are perpendicular to each other so we cannot tie them together.